### ISSUE TITLE
Support for static closures

### ISSUE THREAD

**Issue by xldenis** [xldenis/creusot#34]
I have a _rough_ plan on how we can support static closures. 
Essentially the idea is to make them 'compile time arguments' for the function using them. 
Since we now use modules to represent individual functions, we can make the closure a an opaque `val` inside the function's module, this means that closures can also have contracts! effectively hacking limited higher-order into why3. 
Each time a function is called, we would need to be able to statically determine which closure to clone the function with, which should not be an issue for non-dynamic closures. 

So to recap with an example, given:

```rust
x.map(|x| { ... })
```

The closure will be given an anonymous type (and name) by rustc, which we use to compile and call map in the following manner:

```
module Map
  type t
  type u
  val closure0 : t -> u
  ...
end

...

clone Map with val closure0 = AnonName0.closure, ...

```

**Comment by xldenis** [comment]
Discussed this in person today. It should more or less just come for free from trait support. The only specific point of contention would be to properly translate closure types themselves. (ie for closure construction).
